mm2src/coins/utxo/utxo_tests.rs

```diff
@@ -1196,19 +1195,162 @@ fn test_generate_transaction_relay_fee_is_used_when_dynamic_fee_is_lower() {
+        let mut outputs = vec![];
+        let mut total_outputs = 0_u64;
+        for val in &output_vals {
+            outputs.push(TransactionOutput {
+                script_pubkey: "76a914124b0846223ef78130b8e544b9afc3b09988238688ac".into(),
+                value: *val,
+            });
+            if *val < dust {
+                has_dust_output = true;
+            }
+            total_outputs += *val;
+        }
+
+        let builder = block_on(UtxoTxBuilder::new(&coin))
+            .add_available_inputs(unspents)
+            .add_outputs(outputs.clone())
+            .with_dust(dust)
+            .with_fee(ActualFeeRate::Dynamic(fee_rate));
+
+        let result = block_on(builder.build());
+        if has_dust_output {
+            let is_err_dust = matches!(
+                result.unwrap_err().get_inner(),
+                GenerateTxError::OutputValueLessThanDust { value: _, dust: _ }
+            );
+            assert!(is_err_dust);
+            continue;
+        }
+        if let Err(ref err) = result {
+            let tx_size_max = est_tx_size(input_vals.len(), output_vals.len() + 1);
+            let tx_fee_max = fee_rate * tx_size_max / 1000;
+            if matches!(err.get_inner(), GenerateTxError::NotEnoughUtxos {
+                sum_utxos: _,
+                required: _
+            }) {
+                assert!(total_inputs < total_outputs + tx_fee_max);
+                continue;
+            }
+            panic!("unexpected utxo builder err");
+        }
+
+        let generated = result.unwrap();
+
+        // generated transaction has no change output but dust
+        assert!(generated.0.outputs.len() >= output_vals.len() && generated.0.outputs.len() <= output_vals.len() + 1);
+        let fact_inputs = generated.0.inputs.iter().fold(0u64, |acc, input| acc + input.amount);
+        // total w/o change:
+        let fact_outputs_no_change = generated
+            .0
+            .outputs
+            .iter()
+            .take(output_vals.len())
+            .fold(0u64, |acc, output| acc + output.value);
+
+        assert_eq!(generated.1.spent_by_me, fact_inputs);
+
+        assert_eq!(total_outputs, fact_outputs_no_change);
+
+        assert_eq!(
+            generated.1.spent_by_me,
+            generated.1.fee_amount + generated.1.received_by_me + total_outputs
+        );
+
+        let tx_size = est_tx_size(generated.0.inputs.len(), generated.0.outputs.len());
+        let estimated_txfee = fee_rate * tx_size / 1000;
+        //println!("generated.1.fee_amount={} estimated_txfee={} received_by_me={} output_vals.len={} generated.0.outputs.len={} dust={}, fee_rate={}",
+        //    generated.1.fee_amount, estimated_txfee, generated.1.received_by_me, output_vals.len(), generated.0.outputs.len(), dust, fee_rate);
+
+        const CHANGE_OUTPUT_SIZE: u64 = 1 + 25 + 8;
+        let max_overpay = dust + fee_rate * CHANGE_OUTPUT_SIZE / 1000; // could be slight overpay due to dust change removed from tx
```

Comment:
thanks!
could we include this bound in the tx builder code as well.

Reply:
Should we add this bound to the builder?
What if someone changes the input selection algo and the bound would become different?
And what to do if we hit it anyway, throw an error? (so user won't be able to send a tx).

Reply:
> What if someone changes the input selection algo and the bound would become different?

this is a case where such a bound deems useful. when the input selection algo changes it should still respect such bound for whatever set of input selection.

> And what to do if we hit it anyway, throw an error? (so user won't be able to send a tx).

we should never hit it (unless there is a bug; then it's justified to return an error).

Reply:
> this is a case where such a bound deems useful.
>...
> we should never hit it (unless there is a bug; then it's justified to return an error).

I would say for that tests should be used (and actually we have them)